---
title: "step13_patternAnalysis"
author: "luozhihui"
date: "2019/11/16"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

##use RNA-seq classes see RNA-expression and chip-seq expression value
我们使用RNA的分类结果首先看看RNA的热图

```{r rna expression, include=FALSE}
library(pheatmap)
#df_all = read.delim("/home/zhihl/Project/CRC/RNA_analysis/all_diff_data.txt",sep = "\t", header=T, row.names=1)
df_all = read.delim("/home/zhihl/Project/CRC/Chip_analysis/dff/version0821/final_result/genes.fpkm_table", sep="\t", header=T, row.names=1, check.names = FALSE)


gene_list_1 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/rna/rna_gene_list_cluster1.txt",sep = "\t", header=T, row.names=1)
gene_list_2 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/rna/rna_gene_list_cluster2.txt",sep = "\t", header=T, row.names=1)
gene_list_3 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/rna/rna_gene_list_cluster3.txt",sep = "\t", header=T, row.names=1)
gene_list_4 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/rna/rna_gene_list_cluster4.txt",sep = "\t", header=T, row.names=1)
gene_list_5 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/rna/rna_gene_list_cluster5.txt",sep = "\t", header=T, row.names=1)
gene_list_6 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/rna/rna_gene_list_cluster6.txt",sep = "\t", header=T, row.names=1)
gene_list_7 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/rna/rna_gene_list_cluster7.txt",sep = "\t", header=T, row.names=1)
gene_list_8 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/rna/rna_gene_list_cluster8.txt",sep = "\t", header=T, row.names=1)
gene_list_9 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/rna/rna_gene_list_cluster9.txt",sep = "\t", header=T, row.names=1)

df_sub = df_all[,1:15]
gene_select_id = c(rownames(gene_list_1),rownames(gene_list_3),rownames(gene_list_4),rownames(gene_list_5),rownames(gene_list_6),rownames(gene_list_7),rownames(gene_list_8),rownames(gene_list_9))

df_sub = df_sub[gene_select_id,]

annotation_row=data.frame(Cluster=c(rep(paste("cluster1 n=", nrow(gene_list_1), ""), nrow(gene_list_1)), rep(paste("cluster3 n=", nrow(gene_list_3), ""), nrow(gene_list_3)), rep(paste("cluster4 n=", nrow(gene_list_4), ""), nrow(gene_list_4)), rep(paste("cluster5 n=", nrow(gene_list_5), ""), nrow(gene_list_5)), rep(paste("cluster6 n=", nrow(gene_list_6), ""), nrow(gene_list_6)), rep(paste("cluster7 n=", nrow(gene_list_7), ""), nrow(gene_list_7)), rep(paste("cluster8 n=", nrow(gene_list_8), ""), nrow(gene_list_8)), rep(paste("cluster9 n=", nrow(gene_list_9), ""), nrow(gene_list_9))), row.names = rownames(df_sub))
annotation_col= data.frame(Group = factor(c(rep("Ctrl", 3), rep("Week2", 3), rep("Week4", 3), rep("Week7", 3), rep("Week10", 3)), levels = c("Ctrl", "Week2", "Week4", "Week7", "Week10")))
row.names(annotation_col) <- colnames(df_sub)

anno_colors = list(a = c())

pdf("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/rna/rna_heatmap.pdf")
pheatmap(df_sub, cluster_rows = F, show_rownames = F, cluster_cols = F, scale = "row",annotation_row = annotation_row,color =colorRampPalette(c("#87CEFA", "white", "#CC2121"))(100), annotation_col = annotation_col)
dev.off()

```


##use deeptools draw the chip-seq picture based one the RNA cluster


```{r chip_seq_FPKM, include=FALSE}
#RNA: 
gene_list_1 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/rna/rna_gene_list_cluster1.txt",sep = "\t", header=T, row.names=1)
gene_list_2 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/rna/rna_gene_list_cluster2.txt",sep = "\t", header=T, row.names=1)
gene_list_3 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/rna/rna_gene_list_cluster3.txt",sep = "\t", header=T, row.names=1)
gene_list_4 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/rna/rna_gene_list_cluster4.txt",sep = "\t", header=T, row.names=1)
gene_list_5 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/rna/rna_gene_list_cluster5.txt",sep = "\t", header=T, row.names=1)
gene_list_6 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/rna/rna_gene_list_cluster6.txt",sep = "\t", header=T, row.names=1)
gene_list_7 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/rna/rna_gene_list_cluster7.txt",sep = "\t", header=T, row.names=1)
gene_list_8 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/rna/rna_gene_list_cluster8.txt",sep = "\t", header=T, row.names=1)
gene_list_9 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/rna/rna_gene_list_cluster9.txt",sep = "\t", header=T, row.names=1)

#H3K27me3
gene_list_H3K27me3_1 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/chip_v2/H3K27me3/H3K27me3_1_gene_list.txt",sep = "\t", header=T, row.names=1)
gene_list_H3K27me3_2 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/chip_v2/H3K27me3/H3K27me3_2_gene_list.txt",sep = "\t", header=T, row.names=1)
gene_list_H3K27me3_3 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/chip_v2/H3K27me3/H3K27me3_3_gene_list.txt",sep = "\t", header=T, row.names=1)
gene_list_H3K27me3_4 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/chip_v2/H3K27me3/H3K27me3_4_gene_list.txt",sep = "\t", header=T, row.names=1)
gene_list_H3K27me3_5 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/chip_v2/H3K27me3/H3K27me3_5_gene_list.txt",sep = "\t", header=T, row.names=1)
gene_list_H3K27me3_6 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/chip_v2/H3K27me3/H3K27me3_6_gene_list.txt",sep = "\t", header=T, row.names=1)
gene_list_H3K27me3_7 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/chip_v2/H3K27me3/H3K27me3_7_gene_list.txt",sep = "\t", header=T, row.names=1)
gene_list_H3K27me3_8 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/chip_v2/H3K27me3/H3K27me3_8_gene_list.txt",sep = "\t", header=T, row.names=1)
gene_list_H3K27me3_9 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/chip_v2/H3K27me3/H3K27me3_9_gene_list.txt",sep = "\t", header=T, row.names=1)

#heterochromatin
gene_list_heterochromatin_1 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/chip_v2/heterochromatin/heterochromatin_1_gene_list.txt",sep = "\t", header=T, row.names=1)
gene_list_heterochromatin_2 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/chip_v2/heterochromatin/heterochromatin_2_gene_list.txt",sep = "\t", header=T, row.names=1)
gene_list_heterochromatin_3 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/chip_v2/heterochromatin/heterochromatin_3_gene_list.txt",sep = "\t", header=T, row.names=1)
gene_list_heterochromatin_4 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/chip_v2/heterochromatin/heterochromatin_4_gene_list.txt",sep = "\t", header=T, row.names=1)
gene_list_heterochromatin_5 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/chip_v2/heterochromatin/heterochromatin_5_gene_list.txt",sep = "\t", header=T, row.names=1)
gene_list_heterochromatin_6 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/chip_v2/heterochromatin/heterochromatin_6_gene_list.txt",sep = "\t", header=T, row.names=1)
gene_list_heterochromatin_7 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/chip_v2/heterochromatin/heterochromatin_7_gene_list.txt",sep = "\t", header=T, row.names=1)
gene_list_heterochromatin_8 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/chip_v2/heterochromatin/heterochromatin_8_gene_list.txt",sep = "\t", header=T, row.names=1)
gene_list_heterochromatin_9 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/chip_v2/heterochromatin/heterochromatin_9_gene_list.txt",sep = "\t", header=T, row.names=1)

#enhancer
gene_list_enhancer_1 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/chip_v2/enhancer/enhancer_1_gene_list.txt",sep = "\t", header=T, row.names=1)
gene_list_enhancer_2 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/chip_v2/enhancer/enhancer_2_gene_list.txt",sep = "\t", header=T, row.names=1)
gene_list_enhancer_3 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/chip_v2/enhancer/enhancer_3_gene_list.txt",sep = "\t", header=T, row.names=1)
gene_list_enhancer_4 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/chip_v2/enhancer/enhancer_4_gene_list.txt",sep = "\t", header=T, row.names=1)
gene_list_enhancer_5 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/chip_v2/enhancer/enhancer_5_gene_list.txt",sep = "\t", header=T, row.names=1)
gene_list_enhancer_6 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/chip_v2/enhancer/enhancer_6_gene_list.txt",sep = "\t", header=T, row.names=1)
gene_list_enhancer_7 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/chip_v2/enhancer/enhancer_7_gene_list.txt",sep = "\t", header=T, row.names=1)
gene_list_enhancer_8 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/chip_v2/enhancer/enhancer_8_gene_list.txt",sep = "\t", header=T, row.names=1)
gene_list_enhancer_9 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/chip_v2/enhancer/enhancer_9_gene_list.txt",sep = "\t", header=T, row.names=1)

#H3K4me3
gene_list_H3K4me3_1 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/chip_v2/H3K4me3/H3K4me3_1_gene_list.txt",sep = "\t", header=T, row.names=1)
gene_list_H3K4me3_2 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/chip_v2/H3K4me3/H3K4me3_2_gene_list.txt",sep = "\t", header=T, row.names=1)
gene_list_H3K4me3_3 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/chip_v2/H3K4me3/H3K4me3_3_gene_list.txt",sep = "\t", header=T, row.names=1)
gene_list_H3K4me3_4 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/chip_v2/H3K4me3/H3K4me3_4_gene_list.txt",sep = "\t", header=T, row.names=1)
gene_list_H3K4me3_5 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/chip_v2/H3K4me3/H3K4me3_5_gene_list.txt",sep = "\t", header=T, row.names=1)
gene_list_H3K4me3_6 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/chip_v2/H3K4me3/H3K4me3_6_gene_list.txt",sep = "\t", header=T, row.names=1)
gene_list_H3K4me3_7 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/chip_v2/H3K4me3/H3K4me3_7_gene_list.txt",sep = "\t", header=T, row.names=1)
gene_list_H3K4me3_8 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/chip_v2/H3K4me3/H3K4me3_8_gene_list.txt",sep = "\t", header=T, row.names=1)
gene_list_H3K4me3_9 = read.delim("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/chip_v2/H3K4me3/H3K4me3_9_gene_list.txt",sep = "\t", header=T, row.names=1)
rna <- c(gene_list_1,gene_list_2,gene_list_3,gene_list_4,gene_list_5,gene_list_6,gene_list_7,gene_list_8,gene_list_9)

H3K27 <- c(gene_list_H3K27me3_1,gene_list_H3K27me3_2,gene_list_H3K27me3_3,gene_list_H3K27me3_4,gene_list_H3K27me3_5,gene_list_H3K27me3_6,gene_list_H3K27me3_7,gene_list_H3K27me3_8,gene_list_H3K27me3_9)

enhancer <- c(gene_list_enhancer_1,gene_list_enhancer_2,gene_list_enhancer_3,gene_list_enhancer_4,gene_list_enhancer_5,gene_list_enhancer_6,gene_list_enhancer_7,gene_list_enhancer_8,gene_list_enhancer_9)

heter <- c(gene_list_heterochromatin_1,gene_list_heterochromatin_2,gene_list_heterochromatin_3,gene_list_heterochromatin_4,gene_list_heterochromatin_5,gene_list_heterochromatin_6,gene_list_heterochromatin_7,gene_list_heterochromatin_8,gene_list_heterochromatin_9)

H3K4 <- c(gene_list_H3K4me3_1,gene_list_H3K4me3_2,gene_list_H3K4me3_3,gene_list_H3K4me3_4,gene_list_H3K4me3_5,gene_list_H3K4me3_6,gene_list_H3K4me3_7,gene_list_H3K4me3_8,gene_list_H3K4me3_9)

#h3k27
H3K27_df =data.frame(gene_list_H3K27me3_1=0,gene_list_H3K27me3_2=0,gene_list_H3K27me3_3=0,gene_list_H3K27me3_4=0,gene_list_H3K27me3_5=0,gene_list_H3K27me3_6=0,gene_list_H3K27me3_7=0,gene_list_H3K27me3_8=0,gene_list_H3K27me3_9=0)

h3k27 <- purrr::map(1:9,function(a){
           purrr::map(1:9,function(b){bb = H3K27[b]$symbol
                                      aa = rna[a]$symbol
                                      share = length(intersect(aa, bb))
                                      x = c(share, length(aa), 
                                            20000 - length(aa), length(bb))
                                      pValue = phyper(share, length(aa), 
                                                      20000 - length(aa), 
                                                      length(bb), lower.tail = F)
                                      
                                      return(pValue)
                                      })
                                    })

H3K27_p <- data.frame(do.call(rbind,lapply(h3k27, unlist)))
colnames(H3K27_p) <- c("H3K27me3_1","H3K27me3_2","H3K27me3_3","H3K27me3_4","H3K27me3_5","H3K27me3_6","H3K27me3_7","H3K27me3_8","H3K27me3_9")

#enhancer
enhancer_df =data.frame(gene_list_enhancer_1 =0,gene_list_enhancer_2=0,gene_list_enhancer_3=0,gene_list_enhancer_4=0,gene_list_enhancer_5=0,gene_list_enhancer_6=0,gene_list_enhancer_7=0,gene_list_enhancer_8=0,gene_list_enhancer_9=0)
Enhancer <- purrr::map(1:9,function(a){
  purrr::map(1:9,function(b){bb = enhancer[b]$symbol
  aa = rna[a]$symbol
  share = length(intersect(aa, bb))
  x = c(share, length(aa), 
        20000 - length(aa), length(bb))
  pValue = phyper(share, length(aa), 
                  20000 - length(aa), 
                  length(bb), lower.tail = F)
  
  return(pValue)
  })
})

enhancer_p <- data.frame(do.call(rbind,lapply(Enhancer, unlist)))
colnames(enhancer_p) <- c("enhancer_1","enhancer_2","enhancer_3","enhancer_4","enhancer_5","enhancer_6","enhancer_7","enhancer_8","enhancer_9")
#heterochromatin
heterochromatin_df =data.frame(gene_list_heterochromatin_1 =0,gene_list_heterochromatin_2=0,gene_list_heterochromatin_3=0,gene_list_heterochromatin_4=0,gene_list_heterochromatin_5=0,gene_list_heterochromatin_6=0,gene_list_heterochromatin_7=0,gene_list_heterochromatin_8=0,gene_list_heterochromatin_9=0)
Heterochromatin <- purrr::map(1:9,function(a){
                    purrr::map(1:9,function(b){
                      bb = heter[b]$symbol
                      aa = rna[a]$symbol
                      share = length(intersect(aa, bb))
                      x = c(share, length(aa), 
                            20000 - length(aa), length(bb))
                      pValue = phyper(share, length(aa), 
                                      20000 - length(aa), 
                                      length(bb), lower.tail = F)
                      
                      return(pValue)
                      })
                    })

heterochromatin_p <- data.frame(do.call(rbind,lapply(Heterochromatin, unlist)))
colnames(heterochromatin_p)<-c("heterochromatin_1","heterochromatin_2","heterochromatin_3","heterochromatin_4","heterochromatin_5","heterochromatin_6","heterochromatin_7","heterochromatin_8","heterochromatin_9") 

#h3k4
H3K4_df =data.frame(gene_list_H3K4me3_1 =0,gene_list_H3K4me3_2=0,gene_list_H3K4me3_3=0,gene_list_H3K4me3_4=0,gene_list_H3K4me3_5=0,gene_list_H3K4me3_6=0,gene_list_H3K4me3_7=0,gene_list_H3K4me3_8=0,gene_list_H3K4me3_9=0)
h3k4 <- purrr::map(1:9,function(a){
  purrr::map(1:9,function(b){bb = H3K4[b]$symbol
                             aa = rna[a]$symbol
                             share = length(intersect(aa, bb))
                             x = c(share, length(aa), 
                                  20000 - length(aa), length(bb))
                             pValue = phyper(share, length(aa), 
                                             20000 - length(aa), 
                             length(bb), lower.tail = F)
                             
                             return(pValue)
                             })
                          })

H3K4_p <- data.frame(do.call(rbind,lapply(h3k4, unlist)))
colnames(H3K4_p) <- c("H3K4me3_1","H3K4me3_2","H3K4me3_3","H3K4me3_4","H3K4me3_5","H3K4me3_6","H3K4me3_7","H3K4me3_8","H3K4me3_9")

df <- cbind(H3K27_p,enhancer_p,heterochromatin_p,H3K4_p)
rownames(df) <- c("rna_cluster1","rna_cluster2","rna_cluster3","rna_cluster4","rna_cluster5","rna_cluster6","rna_cluster7","rna_cluster8","rna_cluster9")

mydata <- df[c(1,3:9),]
data_final = -log10(mydata)
data_final[data_final > 5] = 5
pheatmap(as.matrix(data_final), color = colorRampPalette(c("white", "firebrick3"))(100), cluster_cols = FALSE)



```
##draw total peak


```{r draw_table, include=FALSE}

#1
path <- "/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/indensity/RNA_cluster1_region_H3K27ac.table"
df = read.delim(path, sep="\t", header=F, check.names = FALSE)


path_4 <- "/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/indensity/RNA_cluster4_region_H3K27ac.table"
df_4 = read.delim(path_4, sep="\t", header=F, check.names = FALSE)


path_5 <- "/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/indensity/RNA_cluster5_region_H3K27ac.table"
df_5 = read.delim(path_5, sep="\t", header=F, check.names = FALSE)

path_6 <- "/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/indensity/RNA_cluster6_region_H3K27ac.table"
df_6 = read.delim(path_6, sep="\t", header=F, check.names = FALSE)

path_7 <- "/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/indensity/RNA_cluster7_region_H3K27ac.table"
df_7 = read.delim(path_7, sep="\t", header=F, check.names = FALSE)

path_8 <- "/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/indensity/RNA_cluster8_region_H3K27ac.table"
df_8 = read.delim(path_8, sep="\t", header=F, check.names = FALSE)

path_9 <- "/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/indensity/RNA_cluster9_region_H3K27ac.table"
df_9 = read.delim(path_9, sep="\t", header=F, check.names = FALSE)


data_df = data.frame(Group=c(rep("ctrl", 1000), rep("2week", 1000), rep("4week", 1000), rep("7week", 1000),rep("10week", 1000),
                               rep("ctrl", 1000), rep("2week", 1000), rep("4week", 1000), rep("7week", 1000),rep("10week", 1000),
                             rep("ctrl", 1000), rep("2week", 1000), rep("4week", 1000), rep("7week", 1000),rep("10week", 1000),
                             rep("ctrl", 1000), rep("2week", 1000), rep("4week", 1000), rep("7week", 1000),rep("10week", 1000),
                             rep("ctrl", 1000), rep("2week", 1000), rep("4week", 1000), rep("7week", 1000),rep("10week", 1000),
                             rep("ctrl", 1000), rep("2week", 1000), rep("4week", 1000), rep("7week", 1000),rep("10week", 1000),
                             rep("ctrl", 1000), rep("2week", 1000), rep("4week", 1000), rep("7week", 1000),rep("10week", 1000)
                               ),
                     Number = c(as.vector(as.matrix(df[3,3:1002])[1,]), as.vector(as.matrix(df[4,3:1002])[1,]), as.vector(as.matrix(df[5,3:1002])[1,]), as.vector(as.matrix(df[6,3:1002])[1,]), as.vector(as.matrix(df[7,3:1002])[1,]), 
                                as.vector(as.matrix(df_4[3,3:1002])[1,]), as.vector(as.matrix(df_4[4,3:1002])[1,]), as.vector(as.matrix(df_4[5,3:1002])[1,]), as.vector(as.matrix(df_4[6,3:1002])[1,]), as.vector(as.matrix(df_4[7,3:1002])[1,]),
                                as.vector(as.matrix(df_5[3,3:1002])[1,]), as.vector(as.matrix(df_5[4,3:1002])[1,]), as.vector(as.matrix(df_5[5,3:1002])[1,]), as.vector(as.matrix(df_5[6,3:1002])[1,]), as.vector(as.matrix(df_5[7,3:1002])[1,]),
                                as.vector(as.matrix(df_6[3,3:1002])[1,]), as.vector(as.matrix(df_6[4,3:1002])[1,]), as.vector(as.matrix(df_6[5,3:1002])[1,]), as.vector(as.matrix(df_6[6,3:1002])[1,]), as.vector(as.matrix(df_6[7,3:1002])[1,]),
                                as.vector(as.matrix(df_7[3,3:1002])[1,]), as.vector(as.matrix(df_7[4,3:1002])[1,]), as.vector(as.matrix(df_7[5,3:1002])[1,]), as.vector(as.matrix(df_7[6,3:1002])[1,]), as.vector(as.matrix(df_7[7,3:1002])[1,]),
                                as.vector(as.matrix(df_8[3,3:1002])[1,]), as.vector(as.matrix(df_8[4,3:1002])[1,]), as.vector(as.matrix(df_8[5,3:1002])[1,]), as.vector(as.matrix(df_8[6,3:1002])[1,]), as.vector(as.matrix(df_8[7,3:1002])[1,]),
                                as.vector(as.matrix(df_9[3,3:1002])[1,]), as.vector(as.matrix(df_9[4,3:1002])[1,]), as.vector(as.matrix(df_9[5,3:1002])[1,]), as.vector(as.matrix(df_9[6,3:1002])[1,]), as.vector(as.matrix(df_9[7,3:1002])[1,])
                                
                                
                                ),
                     Locite = c(rep(1:1000, 35)),
                     Type = c(rep("Cluster1", 5000), rep("Cluster4", 5000), rep("Cluster5", 5000), rep("Cluster6", 5000), rep("Cluster7", 5000), rep("Cluster8", 5000), rep("Cluster9", 5000)    ))


data_df$Number<- as.numeric(as.character(data_df$Number))
library("ggplot2")
library("ggsci")

#pdf("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/indensity/density.pdf")
ggplot(data_df,aes(x=Locite,y=Number,color=Group)) +
  geom_line() +
  scale_color_uchicago()+ 
  theme_bw()+theme(plot.title=element_text(size=rel( 1.5),hjust= 0.5), 
                                      axis.title=element_text(size=rel(1.5)), 
                                      axis.text=element_text(size=rel( 1)), 
                                      panel.grid.major=element_line(color= "white"),
                                      panel.grid.minor=element_line(color= "white"), 
                                      panel.border=element_rect(color= "white"),
                                      axis.line=element_line(color= "black",size= 1)) +
  ylab("Chipseq intensity")+xlab("")+facet_wrap(~Type,  shrink = FALSE, scales = "fixed") +
   scale_x_continuous(breaks = seq(0,1000,by=500), labels=c("-5K", "TSS", "5K"))

#dev.off() 


```



##draw each peak

```{r each, include=FALSE}
path <- "/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/indensity/RNA_cluster1_region_H3K27ac.table"
df = read.delim(path, sep="\t", header=F, check.names = FALSE)

data_df = data.frame(Group=c(rep("ctrl", 1000), rep("2week", 1000), rep("4week", 1000), rep("7week", 1000),rep("10week", 1000)),
                     Number = c(as.vector(as.matrix(df[3,3:1002])[1,]), as.vector(as.matrix(df[4,3:1002])[1,]), as.vector(as.matrix(df[5,3:1002])[1,]),                                   as.vector(as.matrix(df[6,3:1002])[1,]), as.vector(as.matrix(df[7,3:1002])[1,])),
                     Locite = c(rep(1:1000, 35)),
                     Type = c(rep("Cluster1", 1000) ))

data_df$Number<- as.numeric(as.character(data_df$Number))
library("ggplot2")
library("ggsci")

pdf("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/indensity/density.pdf")
ggplot(data_df,aes(x=Locite,y=Number,color=Group)) +
  geom_line() +
  scale_color_uchicago()+ 
  theme_bw()+theme(plot.title=element_text(size=rel( 1.5),hjust= 0.5), 
                                      axis.title=element_text(size=rel(1.5)), 
                                      axis.text=element_text(size=rel( 1)), 
                                      panel.grid.major=element_line(color= "white"),
                                      panel.grid.minor=element_line(color= "white"), 
                                      panel.border=element_rect(color= "white"),
                                      axis.line=element_line(color= "black",size= 1)) +
  ylab("Chipseq intensity")+xlab("")+
   scale_x_continuous(breaks = seq(0,1000,by=500), labels=c("-5K", "TSS", "5K"))

dev.off() 

```


#H3K27ac
```{r final_version, include=FALSE}
path <- "/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/indensity/RNA_cluster1_region_H3K27ac.table"
df = read.delim(path, sep="\t", header=F, check.names = FALSE)

path_2 <- "/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/indensity/RNA_cluster2_region_H3K27ac.table"
df_2 = read.delim(path_2, sep="\t", header=F, check.names = FALSE)

path_3 <- "/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/indensity/RNA_cluster3_region_H3K27ac.table"
df_3 = read.delim(path_3, sep="\t", header=F, check.names = FALSE)


path_4 <- "/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/indensity/RNA_cluster4_region_H3K27ac.table"
df_4 = read.delim(path_4, sep="\t", header=F, check.names = FALSE)


path_5 <- "/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/indensity/RNA_cluster5_region_H3K27ac.table"
df_5 = read.delim(path_5, sep="\t", header=F, check.names = FALSE)

path_6 <- "/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/indensity/RNA_cluster6_region_H3K27ac.table"
df_6 = read.delim(path_6, sep="\t", header=F, check.names = FALSE)

path_7 <- "/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/indensity/RNA_cluster7_region_H3K27ac.table"
df_7 = read.delim(path_7, sep="\t", header=F, check.names = FALSE)

path_8 <- "/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/indensity/RNA_cluster8_region_H3K27ac.table"
df_8 = read.delim(path_8, sep="\t", header=F, check.names = FALSE)

path_9 <- "/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/indensity/RNA_cluster9_region_H3K27ac.table"
df_9 = read.delim(path_9, sep="\t", header=F, check.names = FALSE)


data_df = data.frame(Group=c(rep("ctrl", 1000), rep("2week", 1000), rep("4week", 1000), rep("7week", 1000),rep("10week", 1000),
                               rep("ctrl", 1000), rep("2week", 1000), rep("4week", 1000), rep("7week", 1000),rep("10week", 1000),
                             rep("ctrl", 1000), rep("2week", 1000), rep("4week", 1000), rep("7week", 1000),rep("10week", 1000),
                             rep("ctrl", 1000), rep("2week", 1000), rep("4week", 1000), rep("7week", 1000),rep("10week", 1000),
                             rep("ctrl", 1000), rep("2week", 1000), rep("4week", 1000), rep("7week", 1000),rep("10week", 1000),
                             rep("ctrl", 1000), rep("2week", 1000), rep("4week", 1000), rep("7week", 1000),rep("10week", 1000),
                             rep("ctrl", 1000), rep("2week", 1000), rep("4week", 1000), rep("7week", 1000),rep("10week", 1000),
                             rep("ctrl", 1000), rep("2week", 1000), rep("4week", 1000), rep("7week", 1000),rep("10week", 1000),
                             rep("ctrl", 1000), rep("2week", 1000), rep("4week", 1000), rep("7week", 1000),rep("10week", 1000)
                               ),
                     Number = c(as.vector(as.matrix(df[3,3:1002])[1,]), as.vector(as.matrix(df[4,3:1002])[1,]), as.vector(as.matrix(df[5,3:1002])[1,]), as.vector(as.matrix(df[6,3:1002])[1,]), as.vector(as.matrix(df[7,3:1002])[1,]), 
                                as.vector(as.matrix(df_2[3,3:1002])[1,]), as.vector(as.matrix(df_2[4,3:1002])[1,]), as.vector(as.matrix(df_2[5,3:1002])[1,]), as.vector(as.matrix(df_2[6,3:1002])[1,]), as.vector(as.matrix(df_2[7,3:1002])[1,]),
                                as.vector(as.matrix(df_3[3,3:1002])[1,]), as.vector(as.matrix(df_3[4,3:1002])[1,]), as.vector(as.matrix(df_3[5,3:1002])[1,]), as.vector(as.matrix(df_3[6,3:1002])[1,]), as.vector(as.matrix(df_3[7,3:1002])[1,]),
                                as.vector(as.matrix(df_4[3,3:1002])[1,]), as.vector(as.matrix(df_4[4,3:1002])[1,]), as.vector(as.matrix(df_4[5,3:1002])[1,]), as.vector(as.matrix(df_4[6,3:1002])[1,]), as.vector(as.matrix(df_4[7,3:1002])[1,]),
                                as.vector(as.matrix(df_5[3,3:1002])[1,]), as.vector(as.matrix(df_5[4,3:1002])[1,]), as.vector(as.matrix(df_5[5,3:1002])[1,]), as.vector(as.matrix(df_5[6,3:1002])[1,]), as.vector(as.matrix(df_5[7,3:1002])[1,]),
                                as.vector(as.matrix(df_6[3,3:1002])[1,]), as.vector(as.matrix(df_6[4,3:1002])[1,]), as.vector(as.matrix(df_6[5,3:1002])[1,]), as.vector(as.matrix(df_6[6,3:1002])[1,]), as.vector(as.matrix(df_6[7,3:1002])[1,]),
                                as.vector(as.matrix(df_7[3,3:1002])[1,]), as.vector(as.matrix(df_7[4,3:1002])[1,]), as.vector(as.matrix(df_7[5,3:1002])[1,]), as.vector(as.matrix(df_7[6,3:1002])[1,]), as.vector(as.matrix(df_7[7,3:1002])[1,]),
                                as.vector(as.matrix(df_8[3,3:1002])[1,]), as.vector(as.matrix(df_8[4,3:1002])[1,]), as.vector(as.matrix(df_8[5,3:1002])[1,]), as.vector(as.matrix(df_8[6,3:1002])[1,]), as.vector(as.matrix(df_8[7,3:1002])[1,]),
                                as.vector(as.matrix(df_9[3,3:1002])[1,]), as.vector(as.matrix(df_9[4,3:1002])[1,]), as.vector(as.matrix(df_9[5,3:1002])[1,]), as.vector(as.matrix(df_9[6,3:1002])[1,]), as.vector(as.matrix(df_9[7,3:1002])[1,])

                                ),
                     Locite = c(rep(1:1000, 45)),
                     Type = c(rep("Cluster1", 5000), rep("Cluster2", 5000), rep("Cluster3", 5000), rep("Cluster4", 5000), rep("Cluster5", 5000), rep("Cluster6", 5000), rep("Cluster7", 5000), rep("Cluster8", 5000), rep("Cluster9", 5000) ))

#4


data_df$Number<- as.numeric(as.character(data_df$Number))
library("ggplot2")
library("ggsci")

pdf("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/indensity/H3K27ac_density.pdf")
ggplot(data_df,aes(x=Locite,y=Number,color=Group)) +
  geom_line() +
  scale_color_uchicago()+ 
  theme_bw()+theme(plot.title=element_text(size=rel( 1.5),hjust= 0.5), 
                                      axis.title=element_text(size=rel(1.5)), 
                                      axis.text=element_text(size=rel( 1)), 
                                      panel.grid.major=element_line(color= "white"),
                                      panel.grid.minor=element_line(color= "white"), 
                                      panel.border=element_rect(color= "white"),
                                      axis.line=element_line(color= "black",size= 1)) +
  ylab("Chipseq intensity")+xlab("")+facet_wrap(~Type,scales = "free") +
   scale_x_continuous(breaks = seq(0,1000,by=500), labels=c("-5K", "TSS", "5K"))

dev.off() 
```

##H3K4me1, H3K9me2, H3K9me2,H3K4me3,H3K27me3,
```{r final_version, include=FALSE}
marker = "H3K4me1"

path <- paste("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/indensity/RNA_cluster1_region_", marker, ".table", sep="")
df = read.delim(path, sep="\t", header=F, check.names = FALSE)

path_2 <- paste("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/indensity/RNA_cluster2_region_", marker, ".table", sep="")
df_2 = read.delim(path_2, sep="\t", header=F, check.names = FALSE)

path_3 <- paste("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/indensity/RNA_cluster3_region_", marker, ".table", sep="")
df_3 = read.delim(path_3, sep="\t", header=F, check.names = FALSE)


path_4 <- paste("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/indensity/RNA_cluster4_region_", marker, ".table", sep="")
df_4 = read.delim(path_4, sep="\t", header=F, check.names = FALSE)


path_5 <- paste("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/indensity/RNA_cluster5_region_", marker, ".table", sep="")
df_5 = read.delim(path_5, sep="\t", header=F, check.names = FALSE)

path_6 <- paste("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/indensity/RNA_cluster6_region_", marker, ".table", sep="")
df_6 = read.delim(path_6, sep="\t", header=F, check.names = FALSE)

path_7 <- paste("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/indensity/RNA_cluster7_region_", marker, ".table", sep="")
df_7 = read.delim(path_7, sep="\t", header=F, check.names = FALSE)

path_8 <- paste("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/indensity/RNA_cluster8_region_", marker, ".table", sep="")
df_8 = read.delim(path_8, sep="\t", header=F, check.names = FALSE)

path_9 <- paste("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/indensity/RNA_cluster9_region_", marker, ".table", sep="")
df_9 = read.delim(path_9, sep="\t", header=F, check.names = FALSE)


data_df = data.frame(Group=c(rep("ctrl", 2000), rep("2week", 2000), rep("4week", 2000), rep("7week", 2000),rep("10week", 2000),
                               rep("ctrl", 2000), rep("2week", 2000), rep("4week", 2000), rep("7week", 2000),rep("10week", 2000),
                             rep("ctrl", 2000), rep("2week", 2000), rep("4week", 2000), rep("7week", 2000),rep("10week", 2000),
                             rep("ctrl", 2000), rep("2week", 2000), rep("4week", 2000), rep("7week", 2000),rep("10week", 2000),
                             rep("ctrl", 2000), rep("2week", 2000), rep("4week", 2000), rep("7week", 2000),rep("10week", 2000),
                             rep("ctrl", 2000), rep("2week", 2000), rep("4week", 2000), rep("7week", 2000),rep("10week", 2000),
                             rep("ctrl", 2000), rep("2week", 2000), rep("4week", 2000), rep("7week", 2000),rep("10week", 2000),
                             rep("ctrl", 2000), rep("2week", 2000), rep("4week", 2000), rep("7week", 2000),rep("10week", 2000),
                             rep("ctrl", 2000), rep("2week", 2000), rep("4week", 2000), rep("7week", 2000),rep("10week", 2000)
                  ),
                     Number = c(as.vector(as.matrix(df[3,3:2002])[1,]), as.vector(as.matrix(df[4,3:2002])[1,]), as.vector(as.matrix(df[5,3:2002])[1,]), as.vector(as.matrix(df[6,3:2002])[1,]), as.vector(as.matrix(df[7,3:2002])[1,]), 
                                as.vector(as.matrix(df_2[3,3:2002])[1,]), as.vector(as.matrix(df_2[4,3:2002])[1,]), as.vector(as.matrix(df_2[5,3:2002])[1,]), as.vector(as.matrix(df_2[6,3:2002])[1,]), as.vector(as.matrix(df_2[7,3:2002])[1,]),
                                as.vector(as.matrix(df_3[3,3:2002])[1,]), as.vector(as.matrix(df_3[4,3:2002])[1,]), as.vector(as.matrix(df_3[5,3:2002])[1,]), as.vector(as.matrix(df_3[6,3:2002])[1,]), as.vector(as.matrix(df_3[7,3:2002])[1,]),
                                as.vector(as.matrix(df_4[3,3:2002])[1,]), as.vector(as.matrix(df_4[4,3:2002])[1,]), as.vector(as.matrix(df_4[5,3:2002])[1,]), as.vector(as.matrix(df_4[6,3:2002])[1,]), as.vector(as.matrix(df_4[7,3:2002])[1,]),
                                as.vector(as.matrix(df_5[3,3:2002])[1,]), as.vector(as.matrix(df_5[4,3:2002])[1,]), as.vector(as.matrix(df_5[5,3:2002])[1,]), as.vector(as.matrix(df_5[6,3:2002])[1,]), as.vector(as.matrix(df_5[7,3:2002])[1,]),
                                as.vector(as.matrix(df_6[3,3:2002])[1,]), as.vector(as.matrix(df_6[4,3:2002])[1,]), as.vector(as.matrix(df_6[5,3:2002])[1,]), as.vector(as.matrix(df_6[6,3:2002])[1,]), as.vector(as.matrix(df_6[7,3:2002])[1,]),
                                as.vector(as.matrix(df_7[3,3:2002])[1,]), as.vector(as.matrix(df_7[4,3:2002])[1,]), as.vector(as.matrix(df_7[5,3:2002])[1,]), as.vector(as.matrix(df_7[6,3:2002])[1,]), as.vector(as.matrix(df_7[7,3:2002])[1,]),
                                as.vector(as.matrix(df_8[3,3:2002])[1,]), as.vector(as.matrix(df_8[4,3:2002])[1,]), as.vector(as.matrix(df_8[5,3:2002])[1,]), as.vector(as.matrix(df_8[6,3:2002])[1,]), as.vector(as.matrix(df_8[7,3:2002])[1,]),
                                as.vector(as.matrix(df_9[3,3:2002])[1,]), as.vector(as.matrix(df_9[4,3:2002])[1,]), as.vector(as.matrix(df_9[5,3:2002])[1,]), as.vector(as.matrix(df_9[6,3:2002])[1,]), as.vector(as.matrix(df_9[7,3:2002])[1,])

                                ),
                     Locite = c(rep(1:2000, 45)),
                     Type = c(rep("Cluster1", 10000), rep("Cluster2", 10000), rep("Cluster3", 10000), rep("Cluster4", 10000), rep("Cluster5", 10000), rep("Cluster6", 10000), rep("Cluster7", 10000), rep("Cluster8", 10000), rep("Cluster9", 10000) ))

#4


data_df$Number<- as.numeric(as.character(data_df$Number))
library("ggplot2")
library("ggsci")

pdf(paste("/home/zhihl/Project/CRC/Chip_analysis/MaSigPro/indensity/", marker,"_density.pdf", sep=""))
ggplot(data_df,aes(x=Locite,y=Number,color=Group)) +
  geom_line() +
  scale_color_uchicago()+ 
  theme_bw()+theme(plot.title=element_text(size=rel( 1.5),hjust= 0.5), 
                                      axis.title=element_text(size=rel(1.5)), 
                                      axis.text=element_text(size=rel( 1)), 
                                      panel.grid.major=element_line(color= "white"),
                                      panel.grid.minor=element_line(color= "white"), 
                                      panel.border=element_rect(color= "white"),
                                      axis.line=element_line(color= "black",size= 1)) +
  ylab("Chipseq intensity")+xlab("")+facet_wrap(~Type,scales = "free") +
  scale_x_continuous(breaks = seq(0,2000,by=1000), labels=c("-10K", "TSS", "10K"))

dev.off() 
```